policy/handler.go

```diff
@@ -28,160 +68,283 @@ type Handler struct {
+
+		mg, err := h.pm.GetAPMLooker(check.Source)
+		if err != nil {
+			return fmt.Errorf("failed to get APM for strategy %s: %w", check.Strategy.Name, err)
+		}
+
+		runner := newCheckRunner(&CheckRunnerConfig{
+			Log: h.log.Named("check_handler").With("check", check.Name,
+				"source", check.Source, "strategy", check.Strategy.Name),
+			StrategyRunner: s,
+			MetricsGetter:  mg,
+			Policy:         h.policy,
+		}, check)
+
+		h.checkRunners = append(h.checkRunners, runner)
 	}
 
+	return nil
+}
+
+//	Run starts the handler for the given policy
+//
+// This function blocks until the context provided is canceled.
+func (h *Handler) Run(ctx context.Context) {
 	h.ticker = time.NewTicker(h.policy.EvaluationInterval)
 	defer h.ticker.Stop()
 
 	h.log.Trace("starting policy handler")
+
 	for {
 		select {
 		case <-ctx.Done():
 			h.log.Error("stopping policy handler due to context done")
+			h.errChn <- ctx.Err()
 			return
 
 		case updatedPolicy := <-h.updatesCh:
-			err := updatedPolicy.Validate()
-			if err != nil {
-				h.errChn <- fmt.Errorf("invalid policy: %v", err)
+			// The policy can be nil if the channel is closed meaning the
+			// handler is being removed, let the context cancelation take
+			// care of returning.
+			if updatedPolicy == nil {
+				continue
 			}
 
 			h.applyMutators(updatedPolicy)
 			h.updateHandler(updatedPolicy)
 
 		case <-h.ticker.C:
-			eval, err := h.handleTick(ctx, h.policy)
+
+			currentStatus, err := h.runTargetStatus()
 			if err != nil {
-				h.errChn <- fmt.Errorf("handler: unable to process policy %v", err)
+				h.errChn <- fmt.Errorf("handler: failed to get target status for target: %s, %w",
+					h.policy.Target.Name, err)
+			}
+
+			// A nil status indicates the target doesn't exist, log and return.
+			if currentStatus == nil {
+				h.errChn <- errTargetNotFound
+				continue
 			}
 
-			if eval != nil {
-				config.EvalsChannel <- eval
+			if currentStatus != nil && !currentStatus.Ready {
+				h.log.Debug("skipping evaluation, target not ready")
+				continue
 			}
 
-		case ts := <-h.cooldownCh:
-			// Enforce the cooldown which will block until complete.
-			if !h.enforceCooldown(ctx, ts) {
-				// Context was canceled, return to stop the handler.
-				return
+			currentCount := currentStatus.Count
+			action, err := h.CalculateNewCount(ctx, currentCount)
+			if err != nil {
+				h.errChn <- fmt.Errorf("handler: unable to execute policy ID: %s, %w",
+					h.policy.ID, err)
+				continue
+			}
+
+			// Canonicalize action so plugins don't have to.
+			action.Canonicalize()
+			action.CapCount(h.policy.Min, h.policy.Max)
+
+			h.log.Info("calculating scaling target", "from", currentCount, "to",
+				action.Count, "reason", action.Reason, "meta", action.Meta)
+
+			if !scalingNeeded(action, currentCount) {
+				h.log.Info("skipping scaling, no action needed")
+				continue
+			}
+
+			h.UpdateNextAction(action)
+
+			if nowFunc().After(h.OutOfCooldownOn()) {
+				h.UpdateState(StateIdle)
+			}
```

Comment:
I can't find where we're resetting the `outOfCooldownOn` field. Should we make sure we're also checking that we're in `StateCooldown` here and/or clearing `outOfCooldownOn` once we have?

Reply:
The `OutOfCooldownOn` is updated once the scaling is successful: https://github.com/hashicorp/nomad-autoscaler/blob/0941fb2b135a1f5fddc04801ccdc5cbad7e04f34/policy/handler.go#L323
For clarity it makes sense to check that we are in cooldown before changing back to idle, but I cant think of a scenario where it could not be.

Reply:
Oh right. For some reason I was thinking we needed to set it to 0 at some point, but that doesn't make sense. LGTM